zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -366,6 +367,27 @@ where
     }
 }
 
+impl UpdateWith<Option<orchard::ShieldedData>> for Chain {
+    fn update_chain_state_with(&mut self, orchard_shielded_data: &Option<orchard::ShieldedData>) {
+        if let Some(orchard_shielded_data) = orchard_shielded_data {
+            for nullifier in orchard_shielded_data.nullifiers() {
+                self.orchard_nullifiers.insert(*nullifier);
+            }
+        }
+    }
+
+    fn revert_chain_state_with(&mut self, orchard_shielded_data: &Option<orchard::ShieldedData>) {
+        if let Some(orchard_shielded_data) = orchard_shielded_data {
+            for nullifier in orchard_shielded_data.nullifiers() {
+                assert!(
+                    self.orchard_nullifiers.remove(nullifier),
+                    "nullifier must be present if block was"
+                );
+            }
+        }
+    }
+}
+
```

Comment:
Please move the changes in the non-finalized state to a separate PR for #1982.

That way, it's easier to make sure each ticket is complete.